Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,356 @@
+                    lines.append(
+                        "%r: Clean bill of health.\n" % (file_path,)
+                    )
+
+                tabnanny.verbose = 1
+                with captured_stdout() as stdout, captured_stderr() as stderr:
+                    tabnanny.check(tmp_dir)
+                stdout = stdout.getvalue()
+                for line in lines:
+                    self.assertIn(line, stdout)
+                self.assertEqual(stderr.getvalue(), "")
+
+    def test_correct_directory(self):
+        """Directory which contains few error free python source code files."""
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            with TemporaryPyFile(
+                SOURCE_CODES["error_free"], directory=tmp_dir
+            ) as file_path:
+                self.verify_tabnanny_check(tmp_dir)
+
+    def test_when_wrong_indented(self):
+        """A python source code file eligible for raising `IndentationError`."""
+        with TemporaryPyFile(
+            SOURCE_CODES["wrong_indented"]
+        ) as file_path:
+            err = ('unindent does not match any outer indentation level'
+                ' (<tokenize>, line 3)\n')
+            err = "%r: Indentation Error: %s" % (file_path, err)
+            self.verify_tabnanny_check(file_path, err=err)
+
+    def test_when_tokenize_tokenerror(self):
+        """A python source code file eligible for raising 'tokenize.TokenError'.
+        """
+        with TemporaryPyFile(
+            SOURCE_CODES["incomplete_expression"]
+        ) as file_path:
+            err = "('EOF in multi-line statement', (7, 0))\n"
+            err = "%r: Token Error: %s" % (file_path, err)
+            self.verify_tabnanny_check(file_path, err=err)
+
+    def test_when_nannynag_error_verbose(self):
+        """A python source code file eligible for raising `tabnanny.NannyNag`.
+
+        Tests will assert `stdout` after activating `tabnanny.verbose` mode.
+        """
+        with TemporaryPyFile(
+            SOURCE_CODES["nannynag_errored"]
+        ) as file_path:
+            out = "%r: *** Line 3: trouble in tab city! ***\n" % (file_path,)
+            out += "offending line: '\\tprint(\"world\")\\n'\n"
+            out += "indent not equal e.g. at tab size 1\n"
+
+            tabnanny.verbose = 1
+            self.verify_tabnanny_check(file_path, out=out)
+
+    def test_when_nannynag_error(self):
+        """A python source code file eligible for raising `tabnanny.NannyNag`.
+        """
+        with TemporaryPyFile(
+            SOURCE_CODES["nannynag_errored"]
+        ) as file_path:
+            out = "%s 3 '\\tprint(\"world\")\\n'\n" % (file_path,)
+            self.verify_tabnanny_check(file_path, out=out)
+
+    def test_when_no_file(self):
+        """A python file which does not exist actually in system."""
+        file_path = 'no_file.py'
+        err = "%r: I/O Error: [Errno 2] No such file or directory: %r\n" % (
```

Comment:
You don't want to tie yourself to exception messages as they do change between versions. Just care about the right exception being raised.

Reply:
I agree, but the code does not raise an exception reference: https://github.com/python/cpython/blob/master/Lib/tabnanny.py#L99
What way will be better to assert the working of mentioned function?